src/template/source.rs

```diff
@@ -0,0 +1,316 @@
+//! The code for error/diagnostics handling is heavily inspiered by
+//! [rust's](https://github.com/rust-lang/rust) compiler. While some code is adpated for use with
+//! punktf, some of it is also a plain copy of it.
+//!
+//! Specifically from those files:
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/lib.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/analyze_source_file.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_parse/src/parser/diagnostics.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/diagnostic.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/diagnostic_builder.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/emitter.rs
```

Comment:
so you copied from multiple files into one? i think we should split this file up, too.

Reply:
`source.rs` and `diagnostic.rs` each contain some code from 2-3 different rust source files. I think we should not split them according to rusts layout as some of these files would only have 10 lines total, but if we find a way to distribute them into more modules which make sense, i am all for it.

Reply:
I would wait with the spliting until we do the cleanup and commenting were we probably will find a good solution as we have to go through them step by step and have a closer look anyway.

Reply:
ok agree, thats part of the cleanup. maybe add a todo note

Reply:
Will do